Refactor permutation importance method and add per-column permutation importance method

angela97lin opened the pull request:
Closes #2299 

- Also updates `calculate_permutation_importance` to only return dict with mean. Previously, the slow impl would return the std / the feature importances but that was different from the fast impl which only returned the mean. Hence, standardizing to return just the mean, since we don't use the std/feature importances to begin with.

Question for discussion: We no longer use the scikit-learn calculate_permutation_importance method as the "slow" method. Should we still compare outputs? I originally wrote this test and confirmed that it passed, but don't think it's necessary to add to codebase. (It also takes 5 minutes for the test locally 😱). Since we're moving away from the sklearn method, I think its fine.
```


@pytest.mark.parametrize('pipeline_class, parameters', test_cases)
@patch('evalml.pipelines.PipelineBase._supports_fast_permutation_importance', new_callable=PropertyMock)
def test_slow_permutation_importance_matches_sklearn_output(mock_supports_fast_importance, pipeline_class, parameters,
                                                            has_minimal_dependencies, fraud_100):
    if has_minimal_dependencies and pipeline_class == LinearPipelineWithTargetEncoderAndOHE:
        pytest.skip("Skipping test_fast_permutation_importance_matches_sklearn_output for target encoder cause "
                    "dependency not installed.")
    from evalml.demos import load_fraud
    from evalml.objectives import get_objective
    from sklearn.inspection import permutation_importance as sk_permutation_importance

    X, y = fraud_100

    if pipeline_class == LinearPipelineWithTextFeatures:
        X.ww.set_types(logical_types={'provider': 'NaturalLanguage'})

    # Do this to make sure we use the same int as sklearn under the hood
    random_state = np.random.RandomState(0)
    random_seed = random_state.randint(np.iinfo(np.int32).max + 1)

    mock_supports_fast_importance.return_value = False
    parameters['Random Forest Classifier'] = {'n_jobs': 1}
    pipeline = pipeline_class(pipeline_class.component_graph, parameters=parameters)
    pipeline.fit(X, y)

    random_state = np.random.RandomState(0)
    random_seed = random_state.randint(np.iinfo(np.int32).max + 1)
    def sklearn_impl(pipeline, X, y, objective, n_repeats, n_jobs, random_seed):
        X = infer_feature_types(X)
        y = infer_feature_types(y)
        objective = get_objective(objective, return_instance=True)

        def scorer(pipeline, X, y):
            scores = pipeline.score(X, y, objectives=[objective])
            return scores[objective.name] if objective.greater_is_better else -scores[objective.name]
        perm_importance = sk_permutation_importance(pipeline, X, y, n_repeats=n_repeats, scoring=scorer, n_jobs=n_jobs,
                                                    random_state=random_seed)
        mean_perm_importance = perm_importance["importances_mean"]
        feature_names = list(X.columns)
        mean_perm_importance = list(zip(feature_names, mean_perm_importance))
        mean_perm_importance.sort(key=lambda x: x[1], reverse=True)
        return pd.DataFrame(mean_perm_importance, columns=["feature", "importance"])

    sklearn_scores = sklearn_impl(pipeline, X, y, objective='Log Loss Binary', n_repeats=5, n_jobs=None,
                                                   random_seed=0)
    mock_supports_fast_importance.return_value = False
    slow_scores = calculate_permutation_importance(pipeline, X, y, objective='Log Loss Binary',
                                                   random_seed=random_seed)
    pd.testing.assert_frame_equal(sklearn_scores, slow_scores)

```

angela97lin commented on evalml/model_understanding/graphs.py:
```diff
@@ -293,113 +293,9 @@ def graph_roc_curve(y_true, y_pred_proba, custom_class_names=None, title_additio
     return _go.Figure(layout=layout, data=graph_data)
 
 
-def _calculate_permutation_scores_fast(pipeline, precomputed_features, y, objective, col_name,
```
I moved everything to a permutation_importance file since we were starting to add more methods that were just to calculate permutation importance. Left the graphing permutation importance method here, though.

freddyaboulton commented:
@angela97lin What i like about that test is that it makes sure our permutation importance implementation is producing correct results (at least by sklearn standards).  I'm not sure what you have in mind for how the new tests would look, but I think it's worth ensuring we're producing correct values (especially since we're venturing out with our own impl now). We don't have to call sklearn's permutation importance to do that though.

angela97lin commented:
@freddyaboulton I converted the previous test which checked against sklearn's impl to compare the new "slow" vs fast method, as a way of checking for correctness. This test which I added above is to check that the "slow" produces the same as sklearn's... which I think is fine to check for this PR but since this PR is also breaking away from the public sklearn interface, is perhaps not necessary? What do you think? 🤔 
I guess TLDR is, how tied are we to checking for correctness via external library :d

freddyaboulton commented:
Got it! I don't think we have to use sklearn to establish "correctness"!

angela97lin commented:
@freddyaboulton Sounds good!! 😁

bchen1116 reviewed (approved):
Looks good! I left a few nit-pick comments, but the test coverage is solid, and I like how you broke the methods apart!

chukarsten reviewed (approved):
Hey Angela, this looks good.  I think there are a few doctstring adjustments to make.  It also seems that the per-column and base perm importance functions are pretty similar so I was wondering if its possible to reuse the function more or less as-is.  I had grand ideas to try it out, but I got distracted :P. Anyways, I'm sure it's the way it is for a reason, so take it or leave it if you think it's worth exploring.

angela97lin commented on evalml/model_understanding/permutation_importance.py:
```diff
@@ -0,0 +1,181 @@
+        y (pd.Series): The target data
+        objective (str, ObjectiveBase): Objective to score on
+        n_repeats (int): Number of times to permute a feature. Defaults to 5.
+        fast (bool): Whether to use the fast method of calculating the permutation importance or not. Defaults to True.
```
Oooooops

angela97lin commented on evalml/model_understanding/permutation_importance.py:
```diff
@@ -0,0 +1,181 @@
+
+    Arguments:
+        pipeline (PipelineBase or subclass): Fitted pipeline
+        X (pd.DataFrame): The input data used to score and compute permutation importance
```
Nay, I think this is cleanest to just support original features and have `col_name` specify the column in the dataframe (which could just be one-column)

angela97lin commented on evalml/model_understanding/permutation_importance.py:
```diff
@@ -0,0 +1,181 @@
+        if is_classification(pipeline.problem_type):
+            y = pipeline._encode_targets(y)
+        baseline_score = _fast_scorer(pipeline, precomputed_features, X, y, objective)
+        scores = _calculate_permutation_scores_fast(
+            pipeline, precomputed_features, y, objective, col_name, random_seed, n_repeats, _fast_scorer, baseline_score,
+        )
```
@chukarsten 
Hmmm yeah I think there is a bunch of overlap, tldr being the one col version uses `_calculate_permutation_scores_fast` while `calculate_permutation_importance` uses `_fast_permutation_importance` (which calls in parallel `_fast_permutation_importance`)...
- one col expects the precomputed features to be passed in as an arg if fast is used. This is primarily for our own reasons / how we're expecting to use this 😝 
- 
I can see the argument for wrangling and combining the two and then have it run just one column but that assumes that we'll just pass in the one feature as `X` if we want the one column method, and that's not what we want! I *did* refactor some underlying logic though to avoid repeats, lmk what ya think!

chukarsten replied:
It's all good, I expected there would be a good reason I wasn't immediately privy to!
